inference-engine/src/preprocessing/ie_preprocess_gapi_kernels.cpp

```diff
@@ -42,572 +42,529 @@
+};
 
-    if (with_cpu_x86_avx2()) {
-        if (std::is_same<T, uint8_t>::value && chs == 2) {
-            avx::splitRow_8UC2(in, outs[0], outs[1], length);
-            return;
-        }
+}  // namespace
 
-        if (std::is_same<T, uint8_t>::value && chs == 3) {
-            avx::splitRow_8UC3(in, outs[0], outs[1], outs[2], length);
-            return;
-        }
+template <typename typelist>
+bool is_cv_type_in_list(const int type_id) {
+    return type_dispatch<typelist>(type_id, cv_type_id{}, [](...){ return true;}, false);
+}
 
-        if (std::is_same<T, uint8_t>::value && chs == 4) {
-            avx::splitRow_8UC4(in, outs[0], outs[1], outs[2], outs[3], length);
-            return;
-        }
+namespace {
+using merge_supported_types = typelist<uint8_t, int8_t, uint16_t, int16_t, int32_t, float, fp_16_t>;
 
-        if (std::is_same<T, float>::value && chs == 2) {
-            avx::splitRow_32FC2(reinterpret_cast<const float*>(in),
-                                reinterpret_cast<float*>(outs[0]),
-                                reinterpret_cast<float*>(outs[1]),
-                                length);
-            return;
+template<typename T, int chs>
+void mergeRowImpl(scalar_tag, const std::array<const T*, chs>& ins, T* out, const int length) {
+    for (int x = 0; x < length; ++x) {
+        for (int c = 0; c < chs; ++c) {
+            out[chs * x + c] = ins[c][x];
         }
+    }
+}
 
-        if (std::is_same<T, float>::value && chs == 3) {
-            avx::splitRow_32FC3(reinterpret_cast<const float*>(in),
-                                reinterpret_cast<float*>(outs[0]),
-                                reinterpret_cast<float*>(outs[1]),
-                                reinterpret_cast<float*>(outs[2]),
-                                length);
-            return;
-        }
+template<typename isa_tag_t, int chs>
+struct typed_merge_row {
+    using p_f = void (*)(const std::array<const uint8_t*, chs>& ins, uint8_t* out, const int length);
 
-        if (std::is_same<T, float>::value && chs == 4) {
-            avx::splitRow_32FC4(reinterpret_cast<const float*>(in),
-                                reinterpret_cast<float*>(outs[0]),
-                                reinterpret_cast<float*>(outs[1]),
-                                reinterpret_cast<float*>(outs[2]),
-                                reinterpret_cast<float*>(outs[3]),
-                                length);
-            return;
-        }
+    template <typename type>
+    p_f operator()(type_to_type<type> ) {
+        return [](const std::array<const uint8_t*, chs>& ins, uint8_t* out, const int length) {
+            const auto inT = reinterpret_cast<const std::array<const type*, chs>&>(ins);
+            auto outT = reinterpret_cast<type*>(out);
+            scalar_tag t;
+            mergeRowImpl<type, chs>(t, inT, outT, length);
+        };
+    }
+#if MANUAL_SIMD
+    p_f operator()(type_to_type<uint8_t>) {
```

Comment:
@AsyaPronina @rgarnov I think that there is possibility to replace conditional compilation macro the std::enable_if. However I don't have idea how do it better. Could you please give me an advice?